packages/react/src/components/createInlineOverlayComponent.tsx

```diff
@@ -54,65 +55,43 @@ export const createInlineOverlayComponent = <PropType, ElementType>(
-       * Also manually call the onWillPresent
-       * handler if present as setState will
-       * cause the event handlers to be
-       * destroyed and re-created.
-       */
-      this.ref.current?.addEventListener('willPresent', (evt: any) => {
-        this.setState({ isOpen: true });
+      this.ref.current?.addEventListener('ionMount', this.handleIonMount);
+      this.ref.current?.addEventListener('willPresent', this.handleWillPresent);
+      this.ref.current?.addEventListener('didDismiss', this.handleDidDismiss);
+    }
 
-        this.props.onWillPresent && this.props.onWillPresent(evt);
-      });
+    componentDidUpdate(prevProps: IonicReactInternalProps<PropType>) {
+      const node = this.ref.current! as HTMLElement;
+      attachProps(node, this.props, prevProps);
+    }
 
+    componentWillUnmount() {
+      const node = this.ref.current;
       /**
-       * Unmount the inner component.
-       * React will call Node.removeChild
-       * which expects the child to be
-       * a direct descendent of the parent
-       * but due to the presence of
-       * Web Component slots, this is not
-       * always the case. To work around this
-       * we move the inner component to the root
-       * of the Web Component so React can
-       * cleanup properly.
+       * If the overlay is being unmounted, but is still
+       * open, this means the unmount was triggered outside
+       * of the overlay being dismissed.
+       *
+       * This can happen with:
+       * - The parent component being unmounted
+       * - The overlay being conditionally rendered
+       * - A route change (push/pop/replace)
+       *
+       * Unmounting the overlay at this stage should skip
+       * the dismiss lifecycle, including skipping the transition.
+       *
        */
-      this.ref.current?.addEventListener('didDismiss', (evt: any) => {
-        const wrapper = this.wrapperRef.current;
-        const el = this.ref.current;
-
+      if (node && this.state.isOpen) {
         /**
-         * This component might be unmounted already, if the containing
-         * element was removed while the popover was still open. (For
-         * example, if an item contains an inline popover with a button
-         * that removes the item.)
+         * Detach the local event listener than performs the state updates,
+         * before dismissing the overlay, to prevent the callback handlers
+         * executing after the component has been unmounted. This is to
+         * avoid memory leaks.
          */
-        if (wrapper && el) {
-          el.append(wrapper);
-          this.setState({ isOpen: false });
-        }
-
-        this.props.onDidDismiss && this.props.onDidDismiss(evt);
-      });
-    }
-
-    componentDidUpdate(prevProps: IonicReactInternalProps<PropType>) {
-      const node = this.ref.current! as HTMLElement;
-      attachProps(node, this.props, prevProps);
+        node.removeEventListener('didDismiss', this.handleDidDismiss);
```

Comment:
Why do we only remove the `didDismiss` event?

Reply:
The `createInlineOverlayComponent` binds two internal event listeners for `didDismiss` and `willPresent`. Both implementations have an internal state update for:

```ts
this.setState({ isOpen: false }); // `true` for willPresent
```

At this point in the lifecycle of the component the component is unmounting. `willPresent` cannot fire. However, `didDismiss` can fire and cause any user-implemented callback handlers for `onDidDismiss` to run after the React component has unmounted. If this occurs, React will throw an exception in the console about a memory leak. 

This code disconnects the internal event listener manually, before the element can dispatch `didDismiss`.

Reply:
Would `detachProps` handle removing that listener? If so, would it make sense to instead call `detachProps` before `node.remove` and then remove this line?

Reply:
`detachProps` only removes event listeners added by `attachProps` (because we track the events on the element node on a key called `__events`). This specific event listener is manually added within the React class and isn't removed as a result of `detachProps`.